aten/src/ATen/cuda/CUDAEvent.h

```diff
@@ -17,19 +17,36 @@ namespace at { namespace cuda {
 /*
 * CUDAEvents are movable not copyable wrappers around CUDA's events.
 *
-* CUDAEvents are constructed lazily when recorded on streams. The events
-* have a device, and this device is acquired from the first recording stream.
-* Later streams that record to the event must share this device, but streams
-* on any device can wait on the event.
+* CUDAEvents are constructed lazily when recorded unless it is reconstructed
+* from a cudaIpcEventHandle_t. The events have a device, and this device is
+* acquired from the first recording stream or passed in when reconstructed from
+* a handle. Later streams that record to the event must share this device, but
+* streams on any device can wait on the event.
 */
 struct AT_CUDA_API CUDAEvent {
   // Constants
   static constexpr unsigned int DEFAULT_FLAGS = cudaEventDisableTiming;
 
   // Constructors
-  CUDAEvent(unsigned int flags = DEFAULT_FLAGS)
+  explicit CUDAEvent(unsigned int flags = DEFAULT_FLAGS)
   : flags_{flags} { }
 
+  // Note: the original event and the reconstructed event now share recorded
+  // activities. Users need to make sure the last recording event (either
+  // original or reconstructed) must not be destructed when synchronize() is
+  // called. Otherwise, the behavior is undefined.
+  explicit CUDAEvent(const cudaIpcEventHandle_t* handle) {
```

Comment:
I think the signature should be:

```
CUDAEvent(Device device, const cudaIpcEventHandle_t* handle);
```

It's better to be explicit about the device (instead of assuming the current device) because it's important that the event be reconstructed on the same device that it was created on.

(I'm ambivalent about the choice of Device vs. DeviceIndex, but I think Device will work  better with our other APIs).

Reply:
I hesitated a bit on this API change. It feels a little inconsistent with `Event.record()` where it implicitly getting the device from the current stream. If we make the new API explicit, do we also want to enforce existing API to be explicit?

Reply:
The difference is that the device here is required information. The exported event already has a device, it's just not encoded in a visible way in cudaIpcEventHandle_t. The full information to import an event is *both* the device and the cudaIpcEventHandle_t.

Reply:
hmm, I am not sure if this is just luck or guaranteed behavior. The `test_event_multi_gpu` test [here](https://github.com/pytorch/pytorch/pull/15974) seems to suggest that an event can be reconstructed from handle on any device (does not have to be the same as the exported event). Let me run some more test.